Source/Core/Common/SPSCQueue.h

```diff
@@ -6,104 +6,181 @@
 // a simple lockless thread-safe,
 // single producer, single consumer queue
 
-#include <algorithm>
 #include <atomic>
-#include <cstddef>
+#include <cassert>
+#include <memory>
 
-#include "Common/CommonTypes.h"
+#if defined(__APPLE__)
+#include <condition_variable>
+// std::atomic<T>::notify_one is "unavailable: introduced in macOS 11.0"
+// TODO: Remove this when macOS required version is bumped to 11.0+
+#define ATOMIC_NOTIFY_ONE_IS_UNAVAILABLE
+#endif
 
 namespace Common
 {
-template <typename T, bool NeedSize = true>
-class SPSCQueue
+namespace detail
+{
+class AtomicSize
 {
 public:
-  SPSCQueue() : m_size(0) { m_write_ptr = m_read_ptr = new ElementPtr(); }
-  ~SPSCQueue()
+  std::size_t Size() const { return m_size.load(std::memory_order_acquire); }
+
+protected:
+  void IncSize() { m_size.fetch_add(1, std::memory_order_release); }
+  void DecSize() { m_size.fetch_sub(1, std::memory_order_release); }
+
+  std::atomic<std::size_t> m_size = 0;
+};
+
+#if defined(ATOMIC_NOTIFY_ONE_IS_UNAVAILABLE)
+class WaitableAtomicSize : public AtomicSize
+{
+public:
+  void WaitForEmpty()
   {
-    // this will empty out the whole queue
-    delete m_read_ptr;
+    std::unique_lock lg{m_lock};
+    m_cond.wait(lg, [this] { return Size() == 0; });
+  }
+
+  void WaitForData()
+  {
+    std::unique_lock lg{m_lock};
+    m_cond.wait(lg, [this] { return Size() != 0; });
+  }
+
+protected:
+  void IncSize()
+  {
+    std::unique_lock lg{m_lock};
+    AtomicSize::IncSize();
+    m_cond.notify_one();
   }
 
-  u32 Size() const
+  void DecSize()
+  {
+    std::unique_lock lg{m_lock};
+    AtomicSize::DecSize();
+    m_cond.notify_one();
+  }
+
+private:
+  std::condition_variable m_cond;
+  std::mutex m_lock;
+};
+#else
+class WaitableAtomicSize : public AtomicSize
+{
+public:
+  void WaitForEmpty()
+  {
+    while (auto const old_size = Size())
+      m_size.wait(old_size, std::memory_order_acquire);
+  }
+
+  void WaitForData() { m_size.wait(0, std::memory_order_acquire); }
```

Comment:
These are only safe from the consumer thread, right? Probably should mention that in a comment.

Reply:
`WaitForEmpty` is for the producer to use.
`WaitForData` is for the consumer to use.

Both are technically "safe" from either thread, in that `WaitForEmpty` can be called from the consumer, and if the queue is empty, it will immediately return, but if it's not empty, it will wait indefinitely because the only thread to make it not be empty is itself. Similar situation for `WaitForData`. But I can add comments.